### DISCUSSION TITLE
Realtime keeping the connection Alive with .net

### DISCUSSION THREAD

**Discussion by MikeCLX** [pocketbase/pocketbase#5746] [Q&A] [answered]
im using realtime with .net and its working great until the PB drops the connection when there is no updates after the 5 mins or so
i send http post messages to api/realtime to do the subscriptions and the changes come back via a callback.

there does not seem to be a callback when the server disconnects. what is the best way to tell the server im still there and to maintain the connection or detect that the server has disconnected? what would be the best way to reconnect? 

I have a windows app that will run 24/7 and want a robust way of knowing that when a record in a collection changes (which might not happen for hours) i can pick it up and respond without having to poll every second 24/7. the realtime is working perfect and responsive for this but only untill the server drops the connection.

im running 0.22.18

Sorry if i have not provided enough information please let me know what other info you would require.

Thanks,

Mike

**Comment by ganigeorgiev** [comment]
> there does not seem to be a callback when the server disconnects...

I don't understand what callback do you expect and where it should fire.

Usually disconnect listeners are handled on the client-side and the specific SSE client is responsible to notify you when the connection interrupts. Adding a disconnect callback in the JS and Dart SDKs was requested and it will be added sometime in the near future for those SDKs but you are using .Net and I can't help you with that.

If you want a server-side callback, then you can make use of the Go/JS [Realtime hooks](https://pocketbase.io/docs/go-event-hooks/#realtime-api-hooks).

In any case, note that usually the SSE client should take care to autoreconnect no matter whether the server dropped the connection due to the idle timeout or because of temp no access to the internet.

The only caveat with the reconnect in PocketBase is because we don't have sessions you need to manually resend your subscriptions after receiving the `PB_CONNECT` event (the JS and Dart SDKs handle this transparently).

**Reply by ganigeorgiev** [reply]
As a side-note, if you want to change the default 5min idle timeout you can adjust it via the `OnRealtimeConnectRequest` hook. Here is an example with the JS pb_hooks:

```js
onRealtimeConnectRequest((e) => {
    // idleTimeout is time.Duration, aka. it is in nanosec
    e.idleTimeout = 24 * 3600000000000 // ~24 hours
})
```

But again, the SSE client you use should take care to auto reconnect in case the connection was interrupted for some reason.

**Reply by MikeCLX** [reply]
thanks for the quick reply, i get the PB_Connect event the fist time i connect but i don't get any events after that if there is no changes to the collection. if there is no changes for the 5 mins then i never get another event.

ill dig into it some more in the morning as it seems that the SSE client is not reconnecting.
I was aware from my searches of the ability to extend the timeout but saw that as just delaying the problem, once i get it recovering correctly then i will extend it out. for now it doing it in 5 mins means i see the problem for finding the recovery solution much quicker :)
 thanks for the infor about having to resend the subscriptions, that would have been the next issue once i get past the reconnection issue

**Reply by ganigeorgiev** [reply, marked as answer]
> that the SSE client is not reconnecting

You may have to implement it on your own because at least in JS the default [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) reconnect/shutdown behavior is not consistent across the different browsers.

You can find the related JS SDK code in [RealtimeService.ts#L436](https://github.com/pocketbase/js-sdk/blob/01951f3a42152e72b79ed5709435edf94dca279f/src/services/RealtimeService.ts#L436)

For the Dart SDK we have a minimal custom SSE client implementation in [sse_client.dart#L207](https://github.com/pocketbase/dart-sdk/blob/d7c0c877587befce623608c899be657bd2f710e2/lib/src/sse/sse_client.dart#L207).

**Comment by MikeCLX** [comment]
Thanks,

Just to clarify that I have this all correct.
PB will maintain a Realtime connection as long as there is data to send on that connection at least every timeout period. (I have proved this is the case)
I.E. every time it sends on that connection it will reset the timeout.
If the timeout expires and no data has been sent it will disconnect the client connection. (again, i seem to have proven this)

from what I understand the server has no way of knowing that the client is there or not, so as long as there is new data for that collection that the client subscribed to, at least every timeout period, it will send forever to the dead connection?  
or does it know that the dead client is gone and drop the subscription? 

Regards,

Mike

**Reply by ganigeorgiev** [reply]
For proper implemented clients the server knows when a client disconnects (this should fail under the [realtime.go#L143](https://github.com/pocketbase/pocketbase/blob/develop/apis/realtime.go#L143) check, which will trigger the related defer calls and remove the associated client from the subscribers list).

The idle timeout was added as a very rudimentary measure against buggy clients that may cause forgotten/zombie connections. It behaves essentially like a "ping" because the SSE spec expects clients to perform auto-reconnect in case a previous established connection was interrupted (_as long as the client is interested in maintaining that connection_).

Unfortunately often a closed connection state can have more than 1 cause and because the rules for the auto-reconnect are not strictly defined, this has lead for example to the earlier mentioned subtle differences between various browser implementations and this is why the SDKs have custom reconnect handling for any server error no matter of the response code _(which we may not even have control over if PocketBase is behind a proxy)_.

If you are writing a custom client my recommendation is to reconnect "indefinitely" unless:
- there are no subscriptions
- the client has explicitly invoked the disconnect
- a specific max fail attempts have been reached (this could be optional depending of the type of app you are building)

In the JS and Dart SDKs the connect/disconnect operations are handled transparently and users operate only with subscribe/unsubscribe.
